offchain/dispatcher/src/sender.rs

```diff
@@ -0,0 +1,220 @@
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use anyhow::Result;
+use async_trait::async_trait;
+use contracts::{authority::Authority, history::Claim};
+use eth_tx_manager::{
+    database::FileSystemDatabase as Database,
+    gas_oracle::DefaultGasOracle as GasOracle,
+    manager::Configuration,
+    time::DefaultTime as Time,
+    transaction::{Priority, Transaction, Value},
+    Chain, Error as TransactionManagerError, TransactionManager,
+};
+use rollups_events::RollupsClaim;
+use state_fold_types::{
+    ethabi::Token,
+    ethers::{
+        self,
+        abi::AbiEncode,
+        middleware::SignerMiddleware,
+        providers::{
+            Http, HttpRateLimitRetryPolicy, MockProvider, Provider, RetryClient,
+        },
+        signers::Signer,
+        types::{Address, Bytes, NameOrAddress},
+    },
+};
+use std::sync::Arc;
+use tracing::{instrument, trace, warn};
+use url::Url;
+
+use crate::{config::DispatcherConfig, signer::ConditionalSigner};
+
+// We added this trait for dependency injection and ease of testing.
+#[async_trait]
+pub trait Sender: std::fmt::Debug + Sized {
+    async fn submit_claim(
+        self,
+        dapp_address: Address,
+        rollups_claim: RollupsClaim,
+    ) -> Result<Self>;
+}
+
+struct SubmittableClaim(Address, RollupsClaim);
+
+impl From<SubmittableClaim> for Bytes {
+    fn from(submittable_claim: SubmittableClaim) -> Self {
+        let SubmittableClaim(dapp_address, claim) = submittable_claim;
+        let claim = Claim {
+            epoch_hash: claim.epoch_hash.into_inner(),
+            first_index: claim.first_index,
+            last_index: claim.last_index,
+        };
+        ethers::abi::encode(&[
+            Token::Address(dapp_address),
+            Token::FixedBytes(claim.encode()),
+        ])
+        .into()
+    }
+}
+
+const MAX_RETRIES: u32 = 10;
+const INITIAL_BACKOFF: u64 = 1000;
+
+/// Instantiates the tx-manager calling `new` or `force_new`.
+macro_rules! tx_manager {
+    ($new: ident, $middleware: expr, $database_path: expr, $chain: expr) => {
+        TransactionManager::$new(
+            $middleware.clone(),
+            GasOracle::new(),
+            Database::new($database_path.clone()),
+            $chain,
+            Configuration::default(),
+        )
+        .await
+    };
+}
+
+type Middleware =
+    Arc<SignerMiddleware<Provider<RetryClient<Http>>, ConditionalSigner>>;
+
+type TxManager = TransactionManager<Middleware, GasOracle, Database, Time>;
+
+#[derive(Debug)]
+pub struct ClaimSender {
+    tx_manager: TxManager,
+    confirmations: usize,
+    priority: Priority,
+    from: Address,
+    authority: Authority<Provider<MockProvider>>,
+}
+
+// Private.
+impl ClaimSender {
```

Comment:
The functions in this section are static, so I would remove them from the scope of the claim sender and leave them as private module functions. This is aesthetic, so it is up to you.